### DISCUSSION TITLE
Path Signals and Rail Traffic Control

### DISCUSSION THREAD

**Discussion by lostwizard** [voxeltycoon/issues#472] [General]
**Is your feature request related to a problem? Please describe.**
There are some potential complications with rail signalling in complex layouts that aren't really possible to solve with the current chain and block signals. This could improve throughput in large multi-track station throats or junctions as well as making it easier to prevent the "kissing train" problem.

**Describe the solution you'd like**
A third signal type that works basically like the path signals in OpenTTD. A train approaching a path signal will attempt to reserve a path to its destination from the path signal to the next signal in its path. If it succeeds, it gets a green signal. If it fails, it gets a red signal and has to wait. Reserving the path marks those bits of track as "occupied" and thus will cause any block or chain signals facing into the block to show red (or yellow as appropriate). Perhaps this could be a three head signal that ordinarily reads yellow and switches to red or green for an approaching train.

However, since this is a more complex scheme that probably requires more processing to handle, it seems to make sense to link operation of path signals with the existence of a "rail traffic control" (RTC) system. If the RTC system is not operating, path signals could behave either like block or chain signals (I'm not sure which is more appropriate). If RTC is operating, then the extra path reservation capability would operate. I envision the RTC system not requiring player management, but I would expect it to have an ongoing cost that scales with the number of blocks controlled by at least one path signal to mimic the need to hire traffic controllers. (One suggestion for such would be associating a monthly cost per path signal.) Perhaps there could be a technology requirement for it as well since it would notionally require communication between RTC and the train.

**Describe alternatives you've considered**
Plain path signals without RTC seemed like a good idea, but they have the potential to cause players to simply use them to the exclusion of all else since they simplify signalling quite a bit. (Though maybe a difficulty option to remove the cost for RTC would satisfy casual players who don't want to do complex signal management.)

**Additional context**

**Comment by victorleblais** [comment]
One argument to counter the "players will only use them": I had the same opinion until I heard some people really struggling with signals.

Maybe it's not that bad to have a "solve mostly everything" signals since it could allow players that don't understand / care / want to learn the nuances of signaling to still be able to play with trains. Players that want to take great care of their network will use them only when needed.

**Simple implementation**
Without going too far from the current signal implementation (which the whole RTC would be), simply making them quite expensive (like 10x a block signal) and/or making them have a monthly maintenance cost could show players: "this is not a normal signal, it should be used for special occasions only".

**Another implementation, more involved/complicated**
Make path signal only work if they're within range of a control tower, which could be a small 2 by 1 building.
The range of the control tower could be similar to a warehouse range (maybe in between the "100% profit" and "50% profit" areas?).
The building would cost maybe like 5 or 10k and would have a monthly maintenance cost.
If a path signal is not in range, then it would act like a regular block signal.
The path signal itself would cost the same as a regular signal.


It could look like something like this:
![image](https://user-images.githubusercontent.com/4365075/60762627-271f1400-a065-11e9-901b-cc5b32609c0d.png)

Or this:
![image](https://user-images.githubusercontent.com/4365075/60762632-52a1fe80-a065-11e9-9fe2-e6271c5a4574.png)

**Comment by mayorandrew** [comment]
I think that signals serve two main purposes for every player: avoiding collisions and deadlocks. 
Also more advanced players use signals to boost throughput of their network.

Currently the game has two types of signals: Simple and Chain. They completely fullfill the two main purposes by allowing player to build a network which will not experience deadlocks or collisions.

But in some scenarios player is required to add more track spacing in order to put more Chain signals in order to make blocks smaller and improve throughput. This is where Path signals come in. They allow simple two-way T and X crossings to be much more performant.

I have also though about how Path signals would work in the combination with Chain signals. If train approaches a Chain signal, followed by a Path signal, the path reservation in the Path signal block should happen right when train passes the chain signal. Otherwise there may be a situation where train passes Chain signal but then stops at the Path signal.

Also, if we look at the Transport Fever game, the only signal type available there is the configurable Path signal. It can be one-way or two-way (allowing trains to pass from the back as if there is no signal). With that the game works fine, allowing players to build networks which avoid collisions and deadlocks while having decent throughput. So I think that Path signals would be a welcome addition to the Voxel Tycoon.

**Comment by victorleblais** [comment]
Path signals are planned for Stage 2 of Steam Early Access: https://trello.com/c/QDXuKeaE/6-path-signals.

No details about their implementation has been shared, so I guess that in the meantime, you can post suggestions here to "inspire" the devs :P
